### DISCUSSION TITLE
MODE Verification of Fuzzy Logic Area

### DISCUSSION THREAD

**Discussion by jhicks14** [dtcenter/METplus#1398] [Statistical Computation] [answered]
Hello,

I've been using MODE's fuzzy logic engine to merge my observations, which produces a "new" observation area that outlines the raw observations (see the black outline around the blue gridpoints on the bottom image). 

![image](https://user-images.githubusercontent.com/90361476/152031907-f05209a3-a471-4c53-a16e-a1c9d175fbc8.png)

In this second image, we can see the the fuzzy engine has produced 3 clustered areas out of the initial 5 raw areas. 

![image](https://user-images.githubusercontent.com/90361476/152034331-a65c5784-919a-4928-989e-dba8f7da25d3.png)

I would like to obtain the interest metric of the clustered areas (black outlines), especially since I'd like the three raw objects in the Mid-Atlantic/New England region to be counted as a single merged object. 

Is this information in the text/netcdf files that are produced after running the MODE tool? I've been searching through these files, but I have only come across the interest values for the raw objects. I thought perhaps the "obs_clus_id" field in the netcdf file would contain these new areas, but only the raw object areas are shown (as seen below).

![image](https://user-images.githubusercontent.com/90361476/152048439-de533b0a-b82b-4dcc-874d-703c18b63c42.png)

I also see that the netcdf file contains lat/lon and x/y data for the vertices of the polygons. Are these the vertices for the area created by the fuzzy logic merging? If so, would these just need to be plotted as a shapefile and then read back into METplus for verification?

**Comment by JohnHalleyGotway** [comment]
@jhicks14 I see you have a question about the interest values for pairs of cluster objects from MODE. Good news, the information you're looking for does indeed already exist in the MODE output files. Since I don't have access to the complete output from your example, I'll use a simple example that's generated after compiling MET and running "make test". Here's snapshot of the first page of PostScript output:
<img width="491" alt="Screen Shot 2022-02-02 at 3 05 41 PM" src="https://user-images.githubusercontent.com/21087144/152245313-e6dff96b-599c-49aa-a6bc-a052481e90c9.png">

Notice that there's 1 matched cluster shown in red in this example. The forecast cluster object consists of 2 simple objects and the observation cluster object consists of a single one. Here's a tarfile with the 4 output files created by MODE:
[mode_example.tar.gz](https://github.com/dtcenter/METplus/files/7990376/mode_example.tar.gz)

In particular...
- **mode_120000L_20050807_120000V_120000A.nc** is a NetCDF file containing some potentially useful details.
  - The variables named obs/fcst_clus_hull_lat/lon contain the actually (lat, lon) points of those black outlines around each cluster. You would also need the obs/fcst_clus_hull_start and obs/fcst_clus_hull_npnts values to figure out which (lat, lon)'s go with which cluster. But all that info is there if you actually want it.

- **mode_120000L_20050807_120000V_120000A_obj.txt** is an ascii file with all the summary object information you need.
  - It contains 25 lines: 1 header line followed by 24 data lines. You need to look at the OBJECT_ID column to understand what each line contains.
  - 7 data lines with OBJECT_ID = "F###" followed by 4 data lines with OBJECT_ID = "O###" provide information about the simple forecast and observation objects. These lines describe the location, shape, and intensity of the simple objects.
  - 10 data lines with OBJECT_ID = "F###_O###" provide information about comparing each simple forecast object to each simple observation object. There could be up to 28 of these (7 fcst * 4 obs), but comparisons for objects whose centroids are too far apart are skipped for efficiency. Notice that the INTEREST column is populated in these lines. By default, any pairs with INTEREST > 0.7 are matched.
  - 1 data line with OBJECT_ID = "CF001" and another with "CO001". These lines provide the object location, shape, and intensity for the simple cluster objects (the "C" stands for cluster). So this info is for the group of simple objects that define the cluster. Remember that the forecast cluster is a group of size 2 while the obs is a group of size 1.
  - Finally, the last line is the one you want with OBJECT_ID = "CF001_CO001" which provides information comparing the forecast cluster to its matching observation cluster. And the very last value in the line, has INTEREST = 0.90421.

I want to be clear though that the interest values for the matched clusters does NOT actually impact the processing logic. It is the interest values for the simple object pairs that determine which simple objects from the forecast match which simple objects from the observation. However, MODE recompute the interest value for the matched clusters out of convenience.

Note that it's possible that all simple matched pairs have INTEREST > 0.7 (or whatever threshold you've chosen), but the resulting cluster pairs DO NOT! It's just all depends on the details. Sometimes grouping multiple simple objects together results in a cluster pair with interest < 0.7.

Hope that helps clarify.

**Reply by jhicks14** [reply]
@JohnHalleyGotway Hi John, thanks for the information detailing how to interpret the cluster IDs in the txt/netcdf files!

Perhaps I'm misinterpreting this, but these object IDs are just for the simple objects, correct? I've done some fuzzy logic merging to the observations and forecast objects, which gives me new outlines around each cluster. Should I instead look at the **obs_merge.txt** file that is produced by MET?

Looking at the OBJECT_ID = "CF001_CO001" line in the **obs_merge.txt** file, my interest value looks to be 1, which is not what I'd expect (this time I'm sharing the forecasted and observed fuzzy logic images below).

![mode_cold_hazard_R1_Clustered_Objects_Forecast_FuzzyMerge_20180203](https://user-images.githubusercontent.com/90361476/152258478-b58cc9e6-1fb8-4ad4-a9ab-caada4261297.png)
![mode_cold_hazard_R1_Clustered_Objects_Obs_FuzzyMerge_20180203](https://user-images.githubusercontent.com/90361476/152258479-c136480c-80ab-421b-8cce-a72fef547a5d.png)

The forecast has two clustered objects after fuzzy logic is applied, and the observations have three.  

[mode_cold_hazard_000000L_20180203_000000V_000000A_R1_T1_cts.txt](https://github.com/dtcenter/METplus/files/7990838/mode_cold_hazard_000000L_20180203_000000V_000000A_R1_T1_cts.txt)
[mode_cold_hazard_000000L_20180203_000000V_000000A_R1_T1_fcst_merge.txt](https://github.com/dtcenter/METplus/files/7990839/mode_cold_hazard_000000L_20180203_000000V_000000A_R1_T1_fcst_merge.txt)
[mode_cold_hazard_000000L_20180203_000000V_000000A_R1_T1_obj.txt](https://github.com/dtcenter/METplus/files/7990840/mode_cold_hazard_000000L_20180203_000000V_000000A_R1_T1_obj.txt)
[mode_cold_hazard_000000L_20180203_000000V_000000A_R1_T1_obs_merge.txt](https://github.com/dtcenter/METplus/files/7990841/mode_cold_hazard_000000L_20180203_000000V_000000A_R1_T1_obs_merge.txt)

**Reply by JohnHalleyGotway** [reply]
My apologies. Looks like I dropped the ball on this one.

This [section](https://met.readthedocs.io/en/latest/Users_Guide/mode.html#mode-configuration-file) of the User's Guide describes the MODE configuration options. Please take a look at the "match_flag" and "merge_flag" options in particular. MODE provides 3 different sets of logic users can employ to "merge" objects within the same field.

1. merge_flag provides 2 options...
  - Apply a double-thresholding option for merging.
  - Apply a fuzzy-logic approach for merging by comparing the field to ITSELF.
  - Do both.
2. match_flag provides an additional option...
  - If 2 simple objects in one field MATCH the same simple object in the other, MERGE those 2 objects.

I'm describing these details to point out the "fcst_merge.txt" and "obs_merge.txt" contain the results of comparing each field to ITSELF. Obviously when you compare the same objects to themselves, each object will be a perfect match with itself. But this could be used to capture any nearby objects whose attributes are "close enough" to be merged.

In general, I don't really like this engine merging option.

Personally, I find the double-thresholding and match/merge option to be most intuitive:
```
merge_flag = THRESH;
match_flag = MERGE_BOTH;
```

But you're welcome to configure it however you'd like.

Anywhere you see "F###" or "O###", those indicate simple forecast and observation objects, respectively.
Anywhere you see "CF###" or "CO###", those indicate cluster forecast and observation objects, respectively.

**Reply by jhicks14** [reply]
@JohnHalleyGotway Hi John. Thanks for the info regarding the double threshold merging. I should've been more concise before, but I'm interested primarily in obtaining the convex hull (the black boundary or "rubber band") that surrounds each of the merged clusters, specifically the one produced using the fuzzy engine. In my previous post, I had shown such an image with one of the clusters in the Mid-Atlantic and New England region; this particular cluster was generated only from the fuzzy engine.

I have been able to obtain the convex hull of the clusters generated in the netcdf file (and produce polygons using the x/y points for the convex hull), but I haven't been able to find a way to obtain the convex hull that is produced using the fuzzy logic engine merging.

**Reply by JohnHalleyGotway** [reply]
@jhicks14 thanks for clarifying. It's funny, we actually had a very similar question in discussions recently. Please take a look at this discussion [comment](https://github.com/dtcenter/METplus/discussions/1413#discussioncomment-2113572) explaining how to parse the NetCDF output from MODE. It contains the (lat, lon) points of the boundaries (i.e. outlines) and convex hulls (i.e. rubber bands) for the **simple** fcst and obs objects. It also contains the (lat, lon) points of the convex hulls for the **cluster** fcst and obs objects.

For example, here's the forecast variables (and the obs variables are named similarly) where 'bdy' means boundary and 'hull' means convex hull:
```
	float fcst_simp_bdy_lat(fcst_simp_bdy) ;
	float fcst_simp_bdy_lon(fcst_simp_bdy) ;
	float fcst_simp_hull_lat(fcst_simp_hull) ;
	float fcst_simp_hull_lon(fcst_simp_hull) ;
	float fcst_clus_hull_lat(fcst_clus_hull) ;
	float fcst_clus_hull_lon(fcst_clus_hull) ;
```

Note that this is the output of the entire matching/merging process. The cluster convex hulls points of the fcst-to-fcst and obs-to-obs engine matching step are NOT written to the output NetCDF file. During development, we deemed that level of detail excessive.

If needed, you could of course always run MODE to compare the same forecast field (or observations field) to itself directly. And the resulting NetCDF file MODE output file would contain the data you're requesting.

Hope that helps clarify.

**Reply by jhicks14** [reply]
@JohnHalleyGotway Thanks, John!

I ran MODE to compare the observation fields to themselves, and this gave me a netcdf file with the convex hull vertices I required (and a convex hull that looks the same to the one produced using the fuzzy engine). I then used Python to create filled polygons using these vertices, saved these to a separate netcdf file, and used MODE a second time to verify my forecasts using the new observation objects.

I haven't looked through all of the new object files yet, but I expect there to be a few that have convex hulls that are larger than we expect; i.e., in observation files that have small objects that are spaced apart, one convex hull may be drawn to cluster/merge these two together, even if they should be treated as two separate objects. We'd expect this merging to take place when the objects are close to one another, but I have come across cases where an object in North Dakota is merged with an object in Texas (I'll have to find this image again, but I'll have to sift through quite a few images). Using the fuzzy engine, in this case, would be better at merging only objects that are 'close' to one another, right? Perhaps I'm thinking about this incorrectly.

Should comparing the objects to themselves yield the exact same convex hulls as using the 'engine' option in the 'merge_flag' descriptor? Does the 'engine' option solely change how the interest is calculated (through the use of interest functions and corresponding weights)? How is METplus determining (in both cases) how these convex hulls are drawn, and is there a way to limit how far they are drawn so that objects far from one another aren't clustered?

**Comment by JohnHalleyGotway** [comment]
Justin,

When running MODE to compare forecasts to observations, there are 3 options
for what logic is used to merge objects together. That logic is controlled
by the match_flag and merge_flag configuration options:

https://met.readthedocs.io/en/develop/Users_Guide/mode.html#mode-configuration-file

I want to point out 2 details:

- That merging logic is cumulative. If (1) engine merging and (2) double
threshold merging and (3) merging while matching are all enabled… two
objects showing up as merged in the MODE output may have been merged for
any combination of reasons 1, 2, and 3. There is no easy/obvious way to
disentangle the reasons in the output file.

- A cluster is by definition matched. A cluster is a group of one or more
objects in one field that match a group of one or more objects in the other
field. Let’s say you have a group of objects in the forecast field that are
merged using the engine merging option. If none of the simple objects in
that group match anything in the obs field, then that merged group is
discarded and appear as a set of unmatched simple objects.

Because of the these points, the merging from running MODE fcst vs fcst
will differ from the merging from running MODE fcst vs obs, in general.

I still don’t completely understand what you’re after. If you send some
sample MODE output and images from Python to demonstrate, I could take a
closer look. I still think it’s possible that what you’re looking for
already exists in the output files MODE is already writing.

As for extraneous merges over too large a distance, the fuzzy logic inputs
and weights are all configurable. Users are encouraged to tune them to
produce the type of matching/merging they’d like for the scale of the
feature they’re studying.

John

**Comment by TaraJensen** [comment, marked as answer]
I think this is what Justin is after...

Basically, he wants the area inside the convex hull aka "the black rubber band", not the area of the bitmap mask = 1s inside the hull.  Here's an example of his test data where the hull in the forecast basically includes the contiguous single object and the hull in the obs encompasses several objects and is about the same size as the forecast, but the obs area is a little under 1/3rd of the forecast area.  He couldn't find a measure of the area inside the convex hull so now he's taking the netcdf, filling in all the points in the hull, then passing it back into MODE to get the area.   Is there a way to compute the area of the convex hull without doing this?

![MODE_example_image](https://user-images.githubusercontent.com/23195803/155339894-1c4536ba-3621-4f22-bb76-21fec70f3c3a.png)

**Reply by jhicks14** [reply]
@JohnHalleyGotway Thanks again, John, for clearing things up. I confused the purpose of the MODE interest functions vs the weights. Changing the weights influences how the objects are merged together when using the ENGINE option. I was able to obtain convex hulls that don't try to merge each object across CONUS by changing the weight given to the convex hull distance.

@TaraJensen Thanks for the supplemental information! While I was able to get exactly what we need, like you mentioned, it requires an additional python script that reads in the vertices of the convex hull, fills in these polygons, saves them to a netcdf file, and are then read back into METplus for final verification (once again using MODE).

**Reply by JohnHalleyGotway** [reply]
@jhicks14 thanks for confirming. I'll go ahead and select one of these posts as being the best answer and then lock conversation on this discussion. We've found that when we mark questions as being answered, we often lose track of future posts. So I'll lock it... but please feel free to create a new discussion if more questions/issues arise.
